File: 523 Continuous Subarray Sum/523.py
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        if not nums or len(nums) < 2:
            return False
        sum = 0
        map = {}
        for i in range(len(nums)):
            sum += nums[i]
            sum = sum % k if k != 0 else sum
            if (sum == 0 and i >= 1) or (sum in map and i - map[sum] >= 2):
                return True
            elif sum not in map:
                map[sum] = i
                
        return False

File: 523 Continuous Subarray Sum/523.java
class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        if (nums == null || nums.length < 2) {
            return false;
        }
        int n = nums.length;

        int sum = 0;
        Map<Integer, Integer> prefixSum = new HashMap<>();
        prefixSum.put(0, -1);
        for (int i = 0; i < n; ++i) {
            sum += nums[i];
            sum = k == 0 ? sum : sum % k;
            Integer p = prefixSum.get(sum);
            if (p != null && i - p >= 2) {
                return true;
            }
            if (p == null) {
                prefixSum.put(sum, i);
            }
        }
        return false;
    }

    // Time Complexity: O(n)
    // Space Complexity: O(n)
}